Extractor/Connect/SessionManager.cs

```diff
@@ -0,0 +1,556 @@
+                result.Connection.Session.KeepAlive -= ClientKeepAlive;
+                result.Connection.Session.KeepAlive += ClientKeepAlive;
+                result.Connection.Session.PublishError -= OnPublishError;
+                result.Connection.Session.PublishError += OnPublishError;
+
+                log.LogInformation("Registered new session with ID: {Id}", result.Connection.Session.SessionId);
+
+                if (connection != null)
+                {
+                    await CloseSession(connection.Session, token);
+                }
+                connection = result.Connection;
+            }
+
+            if (result.Connection.Session.SubscriptionCount > 0)
+            {
+                foreach (var sub in result.Connection.Session.Subscriptions)
+                {
+                    log.LogDebug("Session already has subscription: {Sub} with {Count} monitored items", sub.DisplayName, sub.MonitoredItemCount);
+                }
+            }
+
+            state = SessionManagerState.Connected;
+            Context.UpdateFromSession(connection.Session);
+            onStateChange.NotifyAll();
+
+            return true;
+        }
+
+        public async Task CloseSession(ISession toClose, CancellationToken token)
+        {
+            try
+            {
+                var closeTask = toClose.CloseAsync(token);
+                var resultTask = await Task.WhenAny(Task.Delay(5000, token), closeTask);
+                if (closeTask != resultTask)
+                {
+                    log.LogWarning("Failed to close session, timed out");
+                }
+                else
+                {
+                    log.LogInformation("Successfully closed connection to server");
+                }
+            }
+            catch (Exception ex)
+            {
+                log.LogWarning("Failed to close connection to server, proceeding assuming it is closed: {Message}", ex.Message);
+            }
+
+            lock (toClose)
+            {
+                if (!toClose.Disposed)
+                {
+                    toClose.KeepAlive -= ClientKeepAlive;
+                    toClose.PublishError -= OnPublishError;
+                    toClose.Dispose();
+                }
+            }
+        }
+
+        private void OnPublishError(ISession session, PublishErrorEventArgs e)
+        {
+            string symId = StatusCode.LookupSymbolicId(e.Status.Code);
+
+            var sub = session.Subscriptions.FirstOrDefault(sub => sub.Id == e.SubscriptionId);
+
+            if (sub != null)
+            {
+                log.LogError("Unexpected error on publish: {Code}, subscription: {Name}", symId, sub.DisplayName);
+            }
+        }
+
+        private void ClientKeepAlive(ISession sender, KeepAliveEventArgs eventArgs)
+        {
+            client.LogDump("Keep Alive", eventArgs);
+            if (eventArgs.Status == null || !ServiceResult.IsNotGood(eventArgs.Status)) return;
```

Comment:
```suggestion
            if (eventArgs.Status == null || ServiceResult.IsGood(eventArgs.Status)) return;
```

Reply:
Not the same thing! An uncertain keep alive is a failed keep alive.